File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int[] answer = new int[spells.length];
        Arrays.sort(potions);

        for(int i = 0; i < spells.length; i++) {
            int count = potions.length;
            int l = 0;
            int r = potions.length - 1;

            while(l <= r) {
                int mid = (l + r) / 2;

                if((long) spells[i] * potions[mid] >= success) {
                    r = mid - 1;
                    count = mid;
                }
                else {
                    l = mid + 1;
                }
            }
            answer[i] = potions.length - count;
        }


        return answer;
    }
}

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        answer = []
        potions = sorted(potions)

        for i, n in enumerate(spells):
            count = len(potions)
            l = 0
            r = len(potions) - 1

            while l <= r:
                mid = (r + l) // 2

                if n * potions[mid] >= success:
                    count = mid
                    r = mid - 1
                else:
                    l = mid + 1
            
            answer.append(len(potions) - count)
        
        return answer